backgroundfield/backgroundfield.cpp

```diff
@@ -78,10 +80,9 @@ void setBackgroundField(
             
             //Face averages
             for(uint fComponent=0; fComponent<3; fComponent++){
-               bgFunction.setDerivative(0);
-               bgFunction.setComponent((coordinate)fComponent);
+               T3DFunction valueFunction = std::bind(bgFunction, _1, _2, _3, (coordinate)fComponent, 0, (coordinate)0);
```

Comment:
Can you explain to the laymen what this does? :) Can be in Flowdock too.

Reply:
The ```std::bind``` here takes the function that requires 6 arguments, sets the last two to zero, the component is taken from the loop parameter, and the remaining three arguments (x,y,z) are "left open" by the placeholders.
Thus, the return value is a function with three arguments that is assigned to the T3DFunction variable valueFunction.

In principle, the old code did the same thing in an even more cryptic manner. This implementation is not only threadsafe, it is also more transparent to the compiler's optimizer engine.

Reply:
Oh wow this is marvelously weird!

Reply:
I'm not actually fully sure why this was done as an enum instead of a simple int parameter. Maybe for type safety?
Anyway, that part I simply copied over from the old implementation, as all the places in the code where the background field mechanism is actually invoked also pass coordinate-enum values here.